File: DSA_With_C/Algorithms/Sorting/Quick_Sort.c
#include <stdio.h>

void swap(int *val1, int *val2)
{
  int temp = *val1;
  *val1 = *val2;
  *val2 = temp;
}

int partition(int arr[], int low, int high)
{
  int pivot = arr[high];
  int i = (low - 1);
  for (int j = low; j < high; j++)
  {
    if (arr[j] <= pivot)
    {
      i++;
      swap(&arr[i], &arr[j]);
    }
  }
  swap(&arr[i + 1], &arr[high]);
  return (i + 1);
}

void quick_sort(int arr[], int low, int high)
{
  if (low < high)
  {
    int pi = partition(arr, low, high);
    quick_sort(arr, low, pi - 1);
    quick_sort(arr, pi + 1, high);
  }
}

int main()
{
  int arr[] = {5, 2, 4, 6, 1, 3};
  int size = sizeof(arr) / sizeof(arr[0]);
  int low = 0, high = size - 1;

  quick_sort(arr, low, high);

  for (int i = 0; i < size; i++)
  {
    printf("%d ", arr[i]);
  }

  return 0;
}

File: DSA_With_Java/Algorithms/Sorting/QuickSort.java
package DSA_With_Java.Algorithms.Sorting;

public class QuickSort {
  int array[];

  public QuickSort(int array[]) {
    this.array = array;
  }

  void swap(int a, int b) {
    int temp = array[a];
    array[a] = array[b];
    array[b] = temp;
  }

  public int partition(int low, int high) {
    int pivot = array[high];
    int i = low - 1;

    for (int j = low; j <= high - 1; j++) {
      if (array[j] < pivot) {
        i++;
        swap(i, j);
      }
    }

    swap(i + 1, high);
    return (i + 1);
  }

  public void sort(int low, int high) {
    if (low < high) {
      int pivot = partition(low, high);

      sort(low, pivot - 1);
      sort(pivot + 1, high);
    }
  }

  public void display() {
    for (int i = 0; i < array.length; i++) {
      System.out.print(array[i] + " ");
    }
    System.out.println();
  }
}
